### ISSUE TITLE
Dynamically linked dependencies are broken after upgrade

### ISSUE THREAD

**Issue by slhck** [Homebrew/homebrew-core#33364]
- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:

## What you were trying to do (and why)

I wanted to run `ffmpeg` installed through Homebrew. ffmpeg was installed at the time with its default dependencies (including x265) and some others:

```
$ brew info ffmpeg
ffmpeg: stable 4.0.2 (bottled), HEAD
Play, record, convert, and stream audio and video
https://ffmpeg.org/
/usr/local/Cellar/ffmpeg/4.0.2 (283 files, 54.3MB) *
  Built from source on 2018-08-05 at 22:41:05 with: --with-fdk-aac --with-libass --with-libsoxr --with-libssh --with-tesseract --with-libvidstab --with-opencore-amr --with-openh264 --with-openjpeg --with-openssl --with-rtmpdump --with-rubberband --with-sdl2 --with-snappy --with-tools --with-webp --with-x265 --with-xz --with-zeromq --with-zimg --with-fontconfig --with-freetype --with-frei0r --with-libvorbis --with-libvpx --with-opus --with-speex --with-theora --with-two-lame --with-wavpack
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/ffmpeg.rb
==> Dependencies
Build: nasm ✔, pkg-config ✔, texi2html ✔
Required: lame ✔, libvorbis ✔, libvpx ✔, opus ✔, sdl2 ✔, snappy ✔, theora ✔, x264 ✔, x265 ✔, xvid ✔, xz ✔
Optional: chromaprint ✘, fdk-aac ✔, fontconfig ✔, freetype ✔, frei0r ✔, game-music-emu ✘, libass ✔, libbluray ✘, libbs2b ✘, libcaca ✘, libgsm ✘, libmodplug ✘, librsvg ✘, libsoxr ✔, libssh ✘, libvidstab ✔, opencore-amr ✔, openh264 ✔, openjpeg ✔, openssl ✔, rtmpdump ✔, rubberband ✔, speex ✔, srt ✘, tesseract ✔, two-lame ✔, wavpack ✔, webp ✔, zeromq ✔, zimg ✔
```

## What happened (include command output)

ffmpeg aborts due to referencing the wrong x265 dynamic library:

```
$ ffmpeg
dyld: Library not loaded: /usr/local/opt/x265/lib/libx265.160.dylib
  Referenced from: /usr/local/bin/ffmpeg
  Reason: image not found
[1]    41960 abort      ffmpeg
```

## What you expected to happen

I expect ffmpeg (from Homebrew) to be always linked against the correct x265 library (from Homebrew).

This is apparently not the case:

```
$ brew info x265
x265: stable 2.9 (bottled), HEAD
H.265/HEVC encoder
http://x265.org
/usr/local/Cellar/x265/2.8 (11 files, 34.4MB)
  Poured from bottle on 2018-05-24 at 17:13:14
/usr/local/Cellar/x265/2.9 (11 files, 35.2MB) *
  Poured from bottle on 2018-10-22 at 17:15:07
...
```

## Step-by-step reproduction instructions (by running `brew install` commands)

This is a little harder… my assumption is that the upgrade path was as follows:

- `brew install ffmpeg` (requires x265)
- At some point x265 gets updated via Homebrew when doing a `brew upgrade`, but ffmpeg does not get updated during this upgrade
- The linking is broken

This behavior was confirmed by a Homebrew maintainer here: https://twitter.com/c_pellegrino/status/1027896139226525696

## Related links

- https://github.com/Homebrew/homebrew-core/issues/5844 (explains that the formulas can be reinstalled, which of course works, but is not a viable option – IMHO a `brew upgrade` should ensure that the (linked against) dependencies are re-linked.
- https://superuser.com/questions/877510/ffmpeg-dyld-library-not-loaded-usr-local-lib-libmp3lame-0-dylib

**Comment by slhck** [comment]
In the meantime (since I had to get `ffmpeg` working), I did a `brew reinstall ffmpeg`, so I am not at the same state anymore.

I couldn't quite figure out what the steps are to “go back in history” and install a particular, outdated version of a formula, then do the upgrade process.

**Comment by fxcoudert** [comment]
Apparently you had `ffmpeg 4.0.2`, which was not the latest version (it is `4.0.2_1`). I am not sure what state you ended up in, or how you ended up there, but apparently `brew upgrade` failed to upgrade ffmpeg?

It's going to be hard to debug now that everything is gone, though. Happy that you found a solution, closing for now. We'll see if other users are affected.

**Comment by slhck** [comment]
Thanks, but I would really appreciate if this issue was kept open.

The “solution” is more of a workaround; having users run `brew reinstall` only after they've discovered that Homebrew left the system in a broken state is quite troublesome. I've also linked to a Q&A post that describes the same issue, linked to a tweet with Claudia acknowledging that she didn't know Homebrew didn't update these dependencies automatically, and I've had this happen to me several times over the last years.

> Apparently you had ffmpeg 4.0.2, which was not the latest version

Yes, since I haven't run`brew upgrade` in a few weeks, I think.

> apparently brew upgrade failed to upgrade ffmpeg?

Yes and no. Running `brew upgrade` today would have certainly bumped ffmpeg to the newest version, but I'm not upgrading my formulas that often so as not to risk breaking existing scripts (I have to test compatibility first).

The issue is the following — I'm trying to be super explicit:

ffmpeg 4.0.2 was released Jul 18, which is what I installed on Aug 8. It was at the time linked against x265 2.8, which was released May 21 and had been installed on my system since May 25.

x265 2.9 was subsequently released on Oct 8. I installed it on Oct 22 through a normal `brew upgrade`, which then broke ffmpeg, since ffmpeg stayed on the version linked against x265 2.8 (as can be seen in my initial bug report).

Today, did a `brew reinstall ffmpeg`, which of course installed the `revision 1` version of 4.0.2 and fixed the linking. However, ffmpeg's rev 1 was only released on Oct 17—nine days after the bump of x265—which can only mean that anyone who upgraded x265 between Oct 8 and Oct 17 ended up with a broken ffmpeg.

I hope this clarifies what the underlying issue is. And unless I've missed something, I don't think Homebrew reverse-tracks these dependencies yet? In other words, the only workaround I see at the moment is always adding a new `revision` to the “main” formula if a required dependency is updated.

**Comment by slhck** [comment]
All that being said, do you think there is a way to reproduce this by, say, checking out the `homebrew-core` tap at a given point in time, installing `ffmpeg`, then “forwarding” the repo to another state, and running `brew upgrade`?

**Comment by fxcoudert** [comment]
Running `brew upgrade` on Oct 22 should have updated ffmpeg to `4.0.2_1`, which was merged into core on Oct 17. I do not know why it didn't, and I don't see how we can investigate why it did not. *When a library releases an incompatible version, we do update the version of formulas that depend on it, so that they are rebuilt and relinked.* There might be of course cases where we miss it, but it is normally caught by automated testing, so those are rare… and we'll be happy to fix it when they are reported.

However, given that the current version is correct, I am not sure what you expect from leaving this issue open.

**Comment by MikeMcQuaid** [comment]
> I've had this happen to me several times over the last years.

We've very recently improved this functionality. Unless there's anything that we can use to reproduce this it's not actionable enough to leave open, sorry.

**Comment by slhck** [comment]
I completely understand that it's hard to make this actionable. Sorry for not being able to provide more detail. I'll ping back if/when it happens again.

But since you know more about the internals, do you consider it possible that if ffmpeg wasn't upgraded, running `brew cleanup` could have removed the x265 version it was linked against, which would cause this issue?

**Comment by MikeMcQuaid** [comment]
> I completely understand that it's hard to make this actionable. Sorry for not being able to provide more detail. I'll ping back if/when it happens again.

Not your fault, no worries ❤️.

> do you consider it possible that if ffmpeg wasn't upgraded, running `brew cleanup` could have removed the x265 version it was linked against, which would cause this issue?

I don't think it would be but I'm not sure I'm afraid.
